cli/docker/container_manager.go

```diff
@@ -0,0 +1,545 @@
+		errServices := fmt.Errorf("failed to start services")
+		for serviceName, err := range failedServices {
+			errServices = fmt.Errorf("%v: %w", errServices, err)
+			cm.logger.Errorf("Failed to start service container (%s): %s", serviceName, err)
+		}
+		return containers, errServices
+	}
+
+	for _, container := range containers {
+		if err := cm.healthCheckContainer(context.Background(), container); err != nil {
+			return containers, fmt.Errorf("container health check: %w", err)
+		}
+	}
+
+	return containers, nil
+}
+
+func (cm *ContainerManager) GetWorkflowContainer(workflowID string) *RunningContainer {
+	return cm.workflowContainers[workflowID]
+}
+
+func (cm *ContainerManager) GetServiceContainers(workflowID string) []*RunningContainer {
+	return cm.serviceContainers[workflowID]
+}
+
+func (cm *ContainerManager) DestroyAllContainers() error {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	cm.released = true
+	for _, container := range cm.workflowContainers {
+		cm.logger.Infof("ℹ️ Removing workflow container: %s", container.Name)
+		if err := container.Destroy(); err != nil {
+			return fmt.Errorf("destroy workflow container: %w", err)
+		}
+	}
+
+	for _, containers := range cm.serviceContainers {
+		for _, container := range containers {
+			if container == nil {
+				continue
+			}
+			cm.logger.Infof("Removing service container: %s", container.Name)
+			if err := container.Destroy(); err != nil {
+				return fmt.Errorf("destroy service container: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// We are not using the docker sdk for pull, start and create commands because:
+//   - We want to make sure the end user can easily debug using the same docker command we issue
+//     (hard to convert between sdk and cli api)
+//   - We'd like to support options generically,
+//     with the SDK we would need to parse the string ourselves to convert them properly to their own type
+//   - SDK differs from the CLI in some cases, for example pulling from private registry requires the exact token
+//     it can't automatically use the docker config
+func (cm *ContainerManager) runContainer(
+	container models.Container,
+	options containerCreateOptions,
+	envs map[string]string,
+) (*RunningContainer, error) {
+	if cm.released {
+		return nil, fmt.Errorf("container manager was released already")
+	}
+
+	if err := cm.ensureNetwork(); err != nil {
+		return nil, fmt.Errorf("ensure bitrise docker network: %w", err)
+	}
+
+	cm.logger.Infof("ℹ️ Pulling docker image: %s", container.Image)
+	err := cm.pullImageWithRetry(container)
+	if err != nil {
+		return nil, fmt.Errorf("pull docker image: %w", err)
+	}
+	cm.logger.Infof("✅ Docker image pulled: %s", container.Image)
+
+	cm.logger.Infof("ℹ️ Creating docker container: %s", container.Image)
+	err = cm.createContainer(container, options, envs)
+	if err != nil {
+		return nil, fmt.Errorf("create docker container: %w", err)
+	}
+	cm.logger.Infof("✅ Docker container created: %s", container.Image)
```

Comment:
Printing the container ref would be useful

Reply:
Due to the implementation (we are using pure docker commands here) that is not known at this point. The code just below this that tries to start the container will log that information out (as it inspects the running container via the SDK).